src/v/storage/record_batch_builder.h

```diff
@@ -42,15 +45,22 @@ class record_batch_builder {
     struct serialized_record {
         serialized_record(
           iobuf k,
-          iobuf v,
+          std::optional<iobuf> v,
           std::vector<model::record_header> hdrs
           = std::vector<model::record_header>())
           : key(std::move(k))
-          , value(std::move(v))
-          , headers(std::move(hdrs)) {}
+          , headers(std::move(hdrs)) {
+            if (likely(v)) {
+                value = std::move(*v);
+                encoded_value_size = value.size_bytes();
+            } else {
+                encoded_value_size = -1;
+            }
+        }
 
         iobuf key;
         iobuf value;
+        int32_t encoded_value_size;
```

Comment:
very nice. do we want a test for this as well? 

should also this be default initialized with `encoded_value_size{-1}` as well?

Reply:
i can't think of a reason why we'd want a default value since there is no real meaningful default record. anyway this is internal with a non-default constructor.